src/gui/sdlmain.cpp

```diff
@@ -1846,19 +1950,43 @@ static void SetupWindowResolution(const char *val)
 		SDL_Rect bounds;
 		SDL_GetDisplayBounds(sdl.display_number, &bounds);
 		if (w > 0 && h > 0 && w <= bounds.w && h <= bounds.h) {
+			sdl.desktop.want_resizable_window = false;
 			sdl.desktop.window.use_original_size = false;
 			sdl.desktop.window.width = w;
 			sdl.desktop.window.height = h;
 			return;
 		}
 	}
 
-	LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
-	        "'original' instead",
-	        pref.c_str());
+        LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
+                "'original' instead", pref.c_str());
+        sdl.desktop.want_resizable_window = false;
 	sdl.desktop.window.use_original_size = true;
 }
 
+static SDL_Rect CalculateViewport(int win_width, int win_height)
+{
+	const double prog_aspect_ratio = (sdl.draw.width * sdl.draw.scalex) /
+	                                 (sdl.draw.height * sdl.draw.scaley);
+	const double win_aspect_ratio = double(win_width) / double(win_height);
+
+	if (prog_aspect_ratio > win_aspect_ratio) {
+		// match window width
+		const int w = win_width;
+		const int h = win_width / prog_aspect_ratio;
+		const int x = 0;
+		const int y = (win_height - h) / 2;
```

Comment:
(Minor: given these are sizes and dimensions, logically they should never be negative and so we should use unsigned values; however SDL tends to deal in signed values.. for reasons I don't understand). 

`y` could go negative if  `prog_aspect_ratio` is

Reply:
Ah, this caught me by surprise! Thanks for spotting this, it forced me to look at this code again. Thankfully, it will never happen :)

*It would be useful if GitHub markdown had math/logic notation built-in, anyway - let's do it how we did in University times :)*

some premises:
1. `(win_width / win_height) = win_aspect_ratio` =>  
`win_height = (win_width / win_aspect_ratio)`
2. `prog_aspect_ratio > win_aspect_ratio` (in this `if` branch)
3. `(win_width / prog_aspect_ratio) > win_height` 

(1) and (2) and (3) =>  
`(win_width / prog_aspect_ratio) > (win_width / win_aspect_ratio)` =>
`(1 / prog_aspect_ratio) > (1 / win_aspect_ratio)` =>
`prog_aspect_ratio < win_aspect_ratio` (contradiction with 2!)

To be proper numerical proof it should also take into account epsilons to account floating point ops, but I think in this case it wouldn't change the analysis.

This is **very** non-obvious of course, so I added more asserts in code (both to assure code readers and protect if someone will decide to change this code in the future).